### DISCUSSION TITLE
sequence of actions, and arguments, in `app` or `package`

### DISCUSSION THREAD

**Discussion by peterbecich** [srid/haskell-flake#330] [Q&A] [answered]
I have the `hackage-server` Flake using `haskell-flake`: https://github.com/haskell/hackage-server/blob/master/flake.nix

Right now starting the server requires some steps:
https://github.com/haskell/hackage-server?tab=readme-ov-file#nix-run

```
nix run .#hackage-server -- init --static-dir=datafiles

nix run .#hackage-server -- run --static-dir=datafiles --base-uri=http://127.0.0.1:8080
```

How can I write an `app` or `package` to reduce these steps to just `nix run`?

Presently `nix run` fails because I'm not passing any arguments to `hackage-server`:
```
$ nix run
no command given (try --help)
```

I'm aware of `mission-control` but here am trying to make `nix run` do all the work.

Thank you

**Comment by shivaraj-bh** [comment, marked as answer]
This should do the job:

```nix
# Inside `perSystem`
{
  apps.default = {
    type = "app";
    program =
      let
        run-hackage-server = pkgs.writeShellApplication {
          name = "run-hackage-server";
          runtimeInputs = [ config.packages.default ];
          text = ''
            hackage-server init --static-dir=datafiles
            hackage-server run --static-dir=datafiles --base-uri=http://127.0.0.1:8080
          ‘';
        };
      in
      "${lib.getExe run-hackage-server}”;
  };
}
```

**Reply by peterbecich** [reply]
Great, it works, thanks!
https://github.com/haskell/hackage-server/blob/6dbd9eddf173b593a47af205ea43cd051f82e9eb/flake.nix#L18-L54

**Reply by srid** [reply]
Yea. And I believe you don't need `lib.getExe` here.

**Reply by srid** [reply]
`type = "app";` is not required either. This should work:

```nix
{
  apps.default.program = pkgs.writeShellApplication { .. };
}
```

**Reply by peterbecich** [reply]
You're right https://github.com/haskell/hackage-server/blob/bec983da63c867781983cdcb2fad315addc0d041/flake.nix#L18-L42

**Reply by srid** [reply]
By the way, there's also `services-flake` if you are looking for an one-liner way to spin everything up locally.

https://community.flake.parts/services-flake/custom-service
